### PR TITLE
Svtdanny/different samplings

### PR DESCRIPTION
**Pull request by svtdanny**

### CODE REVIEW

File: wrappers/python/nntile/model/generation/llm_samplers.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+import numpy as np
+from scipy.special import softmax
+
+from nntile.model.generation.llm_params import GenerationMode
+
+
+def sample_topk(logits, k, temperature):
+    batch_extended = logits if len(logits.shape) > 1 else logits[:, None]
+    batch_indx = 0
```
Comment (Muxas) [inline wrappers/python/nntile/model/generation/llm_samplers.py:23]: unused variable

File: wrappers/python/nntile/model/generation/llm_samplers.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+from abc import ABC, abstractmethod
+
+import numpy as np
+from scipy.special import softmax
+
+from nntile.model.generation.llm_params import GenerationMode
+
+
+def sample_topk(logits, k, temperature):
+    batch_extended = logits if len(logits.shape) > 1 else logits[:, None]
```
Comment (Muxas) [inline wrappers/python/nntile/model/generation/llm_samplers.py:22]: Further code relies on `len(logits.shape) == 2`, as it accesses `logins.shape[1]`. Maybe just add a corresponding assert? Then, this line will be no needed.

File: wrappers/python/nntile/model/generation/llm_samplers.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        argsorted_logits = np.argsort(single_logits)
+        top_indices = argsorted_logits[-k:]
+        top_probas = softmax(single_logits[top_indices] / temperature, axis=0)
+        next_token = np.random.default_rng().choice(top_indices, p=top_probas)
+        sampled[batch_indx] = next_token
+    return sampled
+
+
+def sample_topp(logits, p_thr, temperature):
+    batch_extended = logits if len(logits.shape) > 1 else logits[:, None]
```
Comment (Muxas) [inline wrappers/python/nntile/model/generation/llm_samplers.py:37]: the same comment, as with `sample_topk()` function

File: wrappers/python/tests/model/generation/test_llm_samplers.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
+# @copyright (c) 2022-present Skolkovo Institute of Science and Technology
+#                              (Skoltech), Russia. All rights reserved.
+#                2023-present Artificial Intelligence Research Institute
+#                              (AIRI), Russia. All rights reserved.
+#
+# NNTile is software framework for fast training of big neural networks on
+# distributed-memory heterogeneous systems based on StarPU runtime system.
+#
+# @file wrappers/python/tests/model/generation/test_llm_samplers.py
+# Test for implementation of Deep ReLU model in NNTile
```
Comment (Muxas) [inline wrappers/python/tests/model/generation/test_llm_samplers.py:10]: unrelated comment

File: wrappers/python/tests/model/generation/test_llm_samplers.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        for _ in range(n_resamples):
+            sampler = TopKSampler(k, temperature)
+            tokens = sampler.sample(inp)
+            sampled.append(tokens)
+
+        sampled = np.vstack(sampled)
+        unique0 = np.unique(sampled[:, 0])
+        unique1 = np.unique(sampled[:, 1])
+
+        assert unique0.sort() == expected0[-k:].sort()
```
Comment (Muxas) [inline wrappers/python/tests/model/generation/test_llm_samplers.py:63]: could you please describe math behind this testing? Some comment is required to justify why this test is statistically correct. It is hard for me to understand it.

File: wrappers/python/tests/model/generation/test_llm_samplers.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            sampled.append(tokens)
+
+        sampled = np.vstack(sampled)
+        unique0 = np.unique(sampled[:, 0])
+        unique1 = np.unique(sampled[:, 1])
+
+        k0 = np.searchsorted(thrs0, p_thr)
+        k1 = np.searchsorted(thrs1, p_thr)
+
+        assert unique0.sort() == expected0[-k0:].sort()
```
Comment (Muxas) [inline wrappers/python/tests/model/generation/test_llm_samplers.py:101]: Hard to understand why this test what it shall do. Need to describe in comments how the testing is meant to do in terms of math.

**Review by Muxas — COMMENTED** [review]
I could not get the testing procedure. Some comments are required to make code clear to me.

File: wrappers/python/nntile/model/generation/llm_samplers.py (hunk 1 above)
Comment (svtdanny) [reply wrappers/python/nntile/model/generation/llm_samplers.py]: removed

File: wrappers/python/nntile/model/generation/llm_samplers.py (hunk 2 above)
Comment (svtdanny) [reply wrappers/python/nntile/model/generation/llm_samplers.py:23]: made asserts
but probably batched and unbatched variations at the same time are good

File: wrappers/python/nntile/model/generation/llm_samplers.py (hunk 3 above)
Comment (svtdanny) [reply wrappers/python/nntile/model/generation/llm_samplers.py:38]: done

File: wrappers/python/tests/model/generation/test_llm_samplers.py (hunk 4 above)
Comment (svtdanny) [reply wrappers/python/tests/model/generation/test_llm_samplers.py]: removed

File: wrappers/python/tests/model/generation/test_llm_samplers.py (hunk 6 above)
Comment (svtdanny) [reply wrappers/python/tests/model/generation/test_llm_samplers.py:125]: add comments to tests

File: wrappers/python/nntile/model/generation/llm_samplers.py (hunk 2 above)
Comment (Muxas) [reply wrappers/python/nntile/model/generation/llm_samplers.py:23]: I meant your code accesses element `logits.shape[1]` which is a runtime error in case `len(logits.shape) == 1`. Assert if len of logits shape is equal 2, not less or equal.

Comment (Muxas) [reply wrappers/python/nntile/model/generation/llm_samplers.py:23]: Alternatively, access `batch_extended.shape[1]` instead of `logits.shape[1]`

File: wrappers/python/nntile/model/generation/llm_samplers.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+from nntile.model.generation.llm_params import GenerationMode
+
+
+def sample_topk(logits, k, temperature):
+    assert len(logits.shape) <= 2, "Support logits or batch of logits only"
+    batch_extended = logits if len(logits.shape) > 1 else logits[:, None]
+
+    sampled = np.empty((batch_extended.shape[1],), dtype=np.int64)
+    for batch_indx in range(logits.shape[1]):
```
Comment (Muxas) [inline wrappers/python/nntile/model/generation/llm_samplers.py:26]: logits.shape[1] -> batch_extended.shape[1]

File: wrappers/python/nntile/model/generation/llm_samplers.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        sampled[batch_indx] = next_token
+    return sampled
+
+
+def sample_topp(logits, p_thr, temperature):
+    assert len(logits.shape) <= 2, "Support logits or batch of logits only"
+    batch_extended = logits if len(logits.shape) > 1 else logits[:, None]
+
+    sampled = np.empty((batch_extended.shape[1],), dtype=np.int64)
+    for batch_indx in range(logits.shape[1]):
```
Comment (Muxas) [inline wrappers/python/nntile/model/generation/llm_samplers.py:41]: logits -> batch_extended

**Review by Muxas — APPROVED** [review]
